perception/traffic_light_arbiter/src/traffic_light_arbiter.cpp

```diff
@@ -0,0 +1,184 @@
+
+#include <lanelet2_extension/utility/message_conversion.hpp>
+#include <rclcpp/time.hpp>
+
+#include <lanelet2_core/primitives/BasicRegulatoryElements.h>
+
+#include <map>
+#include <memory>
+#include <tuple>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace lanelet
+{
+
+using TrafficLightConstPtr = std::shared_ptr<const TrafficLight>;
+
+std::vector<TrafficLightConstPtr> filter_traffic_signals(const LaneletMapConstPtr map)
+{
+  std::vector<TrafficLightConstPtr> signals;
+  for (const auto & element : map->regulatoryElementLayer) {
+    const auto signal = std::dynamic_pointer_cast<const TrafficLight>(element);
+    if (signal) {
+      signals.push_back(signal);
+    }
+  }
+  return signals;
+}
+
+}  // namespace lanelet
+
+TrafficLightArbiter::TrafficLightArbiter(const rclcpp::NodeOptions & options)
+: Node("traffic_light_selector", options)
+{
+  v2x_time_tolerance_ = this->declare_parameter<double>("v2x_time_tolerance", 5.0);
+  perception_time_tolerance_ = this->declare_parameter<double>("perception_time_tolerance", 1.0);
+  v2x_priority_ = this->declare_parameter<bool>("v2x_priority", false);
+
+  map_sub_ = create_subscription<LaneletMapBin>(
+    "~/sub/vector_map", rclcpp::QoS(1).transient_local(),
+    std::bind(&TrafficLightArbiter::onMap, this, std::placeholders::_1));
+
+  perception_tlr_sub_ = create_subscription<TrafficSignalArray>(
+    "~/sub/perception_traffic_signals", rclcpp::QoS(1),
+    std::bind(&TrafficLightArbiter::onPerceptionMsg, this, std::placeholders::_1));
+
+  v2x_tlr_sub_ = create_subscription<TrafficSignalArray>(
+    "~/sub/v2x_traffic_signals", rclcpp::QoS(1),
+    std::bind(&TrafficLightArbiter::onV2xMsg, this, std::placeholders::_1));
+
+  pub_ = create_publisher<TrafficSignalArray>("~/pub/traffic_signals", rclcpp::QoS(1));
+}
+
+void TrafficLightArbiter::onMap(const LaneletMapBin::ConstSharedPtr msg)
+{
+  const auto map = std::make_shared<lanelet::LaneletMap>();
+  lanelet::utils::conversion::fromBinMsg(*msg, map);
+
+  const auto signals = lanelet::filter_traffic_signals(map);
+  map_regulatory_elements_set_.clear();
+  for (const auto & signal : signals) {
+    map_regulatory_elements_set_.emplace(signal->id());
+  }
+}
+
+void TrafficLightArbiter::onPerceptionMsg(const TrafficSignalArray::ConstSharedPtr msg)
+{
+  latest_perception_msg_ = *msg;
+
+  if (
+    (rclcpp::Time(msg->stamp) - rclcpp::Time(latest_v2x_msg_.stamp)).seconds() >
+    v2x_time_tolerance_) {
+    latest_v2x_msg_.signals.clear();
+  }
+
+  arbiterAndPublish(msg->stamp);
+}
+
+void TrafficLightArbiter::onV2xMsg(const TrafficSignalArray::ConstSharedPtr msg)
+{
+  latest_v2x_msg_ = *msg;
+
+  if (
+    (rclcpp::Time(msg->stamp) - rclcpp::Time(latest_perception_msg_.stamp)).seconds() >
+    perception_time_tolerance_) {
+    latest_v2x_msg_.signals.clear();
+  }
+
+  arbiterAndPublish(msg->stamp);
+}
+
+void TrafficLightArbiter::arbiterAndPublish(const builtin_interfaces::msg::Time & stamp)
```

Comment:
[nits]
Maybe verb (arbitrate) is preferred for function name

Reply:
Addressed in [1c2af21](https://github.com/autowarefoundation/autoware.universe/pull/3780/commits/1c2af21cbf6ebe39887f514fa45a19ef0e5b3a07)